src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc

```diff
@@ -218,87 +207,76 @@ void RegisterAppInterfaceRequest::Run() {
     sleep(1);
     // TODO(DK): timer_->StartWait(1);
   }
+}
 
-  if (application_manager_.IsStopping()) {
-    LOG4CXX_WARN(logger_, "The ApplicationManager is stopping!");
-    return;
-  }
+bool RegisterAppInterfaceRequest::IsApplicationForbidden() {
+  const auto& msg_params = (*message_)[strings::msg_params];
+  const std::string policy_app_id = msg_params[strings::app_id].asString();
+  return application_manager_.IsApplicationForbidden(connection_key(),
+                                                     policy_app_id);
+}
 
-  if (IsApplicationSwitched()) {
-    return;
+bool RegisterAppInterfaceRequest::ProcessApplicationTransportSwitching() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  if (!IsApplicationSwitched()) {
+    return false;
   }
+  const auto& msg_params = (*message_)[strings::msg_params];
 
-  const std::string mobile_app_id =
-      (*message_)[strings::msg_params][strings::app_id].asString();
-
-  ApplicationSharedPtr application =
-      application_manager_.application(connection_key());
-
-  if (application) {
+  const auto& policy_app_id = msg_params[strings::app_id].asString();
+  auto app = application_manager_.application_by_policy_id(policy_app_id);
+  DCHECK_OR_RETURN(app, false);
+  if (!application_manager_.IsAppInReconnectMode(policy_app_id)) {
+    LOG4CXX_DEBUG(logger_,
+                  "Policy id " << policy_app_id
+                               << " is not found in reconnection list.");
     SendResponse(false, mobile_apis::Result::APPLICATION_REGISTERED_ALREADY);
-    return;
+    return true;
   }
 
-  const smart_objects::SmartObject& msg_params =
-      (*message_)[strings::msg_params];
-
-  const std::string policy_app_id = msg_params[strings::app_id].asString();
-  std::string new_policy_app_id = policy_app_id;
-  std::transform(policy_app_id.begin(),
-                 policy_app_id.end(),
-                 new_policy_app_id.begin(),
-                 ::tolower);
-  (*message_)[strings::msg_params][strings::app_id] = new_policy_app_id;
+  LOG4CXX_DEBUG(logger_, "Application is found in reconnection list.");
 
-  if (application_manager_.IsApplicationForbidden(connection_key(),
-                                                  policy_app_id)) {
-    SendResponse(false, mobile_apis::Result::TOO_MANY_PENDING_REQUESTS);
-    return;
-  }
+  auto app_type = ApplicationType::kSwitchedApplicationWrongHashId;
+  if ((*message_)[strings::msg_params].keyExists(strings::hash_id)) {
+    const auto hash_id =
+        (*message_)[strings::msg_params][strings::hash_id].asString();
 
-  if (IsApplicationWithSameAppIdRegistered()) {
-    SendResponse(false, mobile_apis::Result::DISALLOWED);
-    return;
+    auto& resume_ctrl = application_manager_.resume_controller();
+    if (resume_ctrl.CheckApplicationHash(app, hash_id)) {
```

Comment:
@LuxoftAKutsan please add some logs to see if hash_id was found and the result of has_id check in resume_ctrl